S3 ETag should not be validated if using KMS SSE

<!--- Provide a general summary of the issue in the Title above -->

## Expected Behavior
[Version 1](https://github.com/aws/aws-sdk-java/blob/5fb7554098a60228347ab823ddde06567f1ef85c/aws-java-sdk-s3/src/main/java/com/amazonaws/services/s3/internal/SkipMd5CheckStrategy.java#L245) of the AWS SDK skipped the ETag validation check if KMS server side encryption is used. This is causing issues using LocalStack which is returning the ETag regardless. 

According to the [API](https://docs.aws.amazon.com/AmazonS3/latest/API/RESTObjectPUT.html)

    In the response, Amazon S3 returns the encryption algorithm and MD5 of the encryption key that you specified when uploading the object. 
    The ETag that is returned is not the MD5 of the object.



## Current Behavior
`ChecksumsEnabledValidator` executes the Checksum check if a ETag is returned, causing a NPE ([as the request context doesn't include the checksum](https://github.com/aws/aws-sdk-java-v2/blob/a0379920f54cba715773f821ac5d237be8a47985/services/s3/src/main/java/software/amazon/awssdk/services/s3/checksums/ChecksumsEnabledValidator.java#L104)).

  
    Caused by: java.lang.NullPointerException: null
    	at software.amazon.awssdk.services.s3.checksums.ChecksumsEnabledValidator.validatePutObjectChecksum(ChecksumsEnabledValidator.java:129)
    	at software.amazon.awssdk.services.s3.internal.handlers.AsyncChecksumValidationInterceptor.afterUnmarshalling(AsyncChecksumValidationInterceptor.java:86)



## Possible Solution
~~Check if the validation should be executed on the response in [ChecksumsEnabledValidator#validatePutObjectChecksum](https://github.com/aws/aws-sdk-java-v2/blob/a0379920f54cba715773f821ac5d237be8a47985/services/s3/src/main/java/software/amazon/awssdk/services/s3/checksums/ChecksumsEnabledValidator.java#L128), similar to V1 or include a null check for `checksum`.~~

## Steps to Reproduce (for bugs)

## Context
Bug raised with https://github.com/localstack/localstack/issues/1092 too.

## Your Environment
* AWS Java SDK version used: 2.3.6
* JDK version used: 1.8
* Operating System and version: Mac OS 10.13.6

--- a/.changes/next-release/bugfix-AWSS3-3112634.json
+++ b/.changes/next-release/bugfix-AWSS3-3112634.json
@@ -1,0 +1,5 @@
+{
+    "category": "AWS S3", 
+    "type": "bugfix", 
+    "description": "Use request header to determine if checksum validation should be enabled for `s3#putObject`"
+}

--- a/services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/handlers/AsyncChecksumValidationInterceptor.java
+++ b/services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/handlers/AsyncChecksumValidationInterceptor.java
@@ -80,7 +80,7 @@ public final class AsyncChecksumValidationInterceptor implements ExecutionInterc
     public void afterUnmarshalling(Context.AfterUnmarshalling context, ExecutionAttributes executionAttributes) {
 
         boolean putObjectChecksumsEnabled =
-            putObjectChecksumEnabled(context.request(), ASYNC, executionAttributes, context.httpResponse());
+            putObjectChecksumEnabled(context.request(), ASYNC, executionAttributes, context.httpRequest());
 
         if (putObjectChecksumsEnabled) {
             validatePutObjectChecksum((PutObjectResponse) context.response(), executionAttributes);

--- a/services/s3/src/test/java/software/amazon/awssdk/services/s3/internal/handlers/AsyncChecksumValidationInterceptorTest.java
+++ b/services/s3/src/test/java/software/amazon/awssdk/services/s3/internal/handlers/AsyncChecksumValidationInterceptorTest.java
@@ -17,30 +17,46 @@ package software.amazon.awssdk.services.s3.internal.handlers;
 
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Java6Assertions.assertThatThrownBy;
 import static software.amazon.awssdk.core.ClientType.ASYNC;
 import static software.amazon.awssdk.core.interceptor.SdkExecutionAttribute.CLIENT_TYPE;
 import static software.amazon.awssdk.core.interceptor.SdkExecutionAttribute.SERVICE_CONFIG;
 import static software.amazon.awssdk.services.s3.checksums.ChecksumConstant.CHECKSUM_ENABLED_RESPONSE_HEADER;
 import static software.amazon.awssdk.services.s3.checksums.ChecksumConstant.CONTENT_LENGTH_HEADER;
 import static software.amazon.awssdk.services.s3.checksums.ChecksumConstant.ENABLE_MD5_CHECKSUM_HEADER_VALUE;
+import static software.amazon.awssdk.services.s3.checksums.ChecksumConstant.SERVER_SIDE_ENCRYPTION_HEADER;
 import static software.amazon.awssdk.services.s3.checksums.ChecksumsEnabledValidator.CHECKSUM;
+import static software.amazon.awssdk.services.s3.model.ServerSideEncryption.AWS_KMS;
 
+import java.net.URI;
 import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
 import java.util.Optional;
 import org.junit.Test;
 import org.reactivestreams.Publisher;
 import software.amazon.awssdk.core.async.AsyncRequestBody;
+import software.amazon.awssdk.core.checksums.Md5Checksum;
+import software.amazon.awssdk.core.checksums.SdkChecksum;
 import software.amazon.awssdk.core.interceptor.Context;
 import software.amazon.awssdk.core.interceptor.ExecutionAttributes;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+import software.amazon.awssdk.http.SdkHttpMethod;
+import software.amazon.awssdk.http.SdkHttpRequest;
 import software.amazon.awssdk.http.SdkHttpResponse;
 import software.amazon.awssdk.services.s3.S3Configuration;
 import software.amazon.awssdk.services.s3.checksums.ChecksumCalculatingAsyncRequestBody;
 import software.amazon.awssdk.services.s3.checksums.ChecksumValidatingPublisher;
 import software.amazon.awssdk.services.s3.model.GetObjectRequest;
 import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.awssdk.services.s3.model.PutObjectResponse;
 import software.amazon.awssdk.services.s3.utils.InterceptorTestUtils;
+import software.amazon.awssdk.utils.internal.Base16Lower;
 
 public class AsyncChecksumValidationInterceptorTest {
+
+    private static final byte[] CONTENT_BYTES = "CONTENT".getBytes(Charset.forName("UTF-8"));
+    private static final String VALID_CHECKSUM = Base16Lower.encodeAsString(checkSumFor(CONTENT_BYTES).getChecksumBytes());
+    private static final String INVALID_CHECKSUM = "3902ee7e149eb8313a34757e89e21af6";
 
     private AsyncChecksumValidationInterceptor interceptor = new AsyncChecksumValidationInterceptor();
 
@@ -115,6 +131,78 @@ public class AsyncChecksumValidationInterceptorTest {
 
     }
 
+    @Test
+    public void afterUnmarshalling_putObjectRequest_shouldValidateChecksum() {
+        SdkHttpResponse sdkHttpResponse = getSdkHttpResponseWithChecksumHeader();
+
+        PutObjectResponse response = PutObjectResponse.builder()
+                                                      .eTag(VALID_CHECKSUM)
+                                                      .build();
+
+        PutObjectRequest putObjectRequest = PutObjectRequest.builder()
+                                                            .build();
+
+        SdkHttpRequest sdkHttpRequest = SdkHttpFullRequest.builder()
+                                                          .uri(URI.create("http://localhost:8080"))
+                                                          .method(SdkHttpMethod.PUT)
+                                                          .build();
+
+        Context.AfterUnmarshalling afterUnmarshallingContext =
+            InterceptorTestUtils.afterUnmarshallingContext(putObjectRequest, sdkHttpRequest, response, sdkHttpResponse);
+
+        interceptor.afterUnmarshalling(afterUnmarshallingContext, getExecutionAttributesWithChecksum());
+    }
+
+    @Test
+    public void afterUnmarshalling_putObjectRequest_shouldValidateChecksum_throwExceptionIfInvalid() {
+        SdkHttpResponse sdkHttpResponse = getSdkHttpResponseWithChecksumHeader();
+
+        PutObjectResponse response = PutObjectResponse.builder()
+                                                      .eTag(INVALID_CHECKSUM)
+                                                      .build();
+
+        PutObjectRequest putObjectRequest = PutObjectRequest.builder().build();
+
+        SdkHttpRequest sdkHttpRequest = SdkHttpFullRequest.builder()
+                                                          .uri(URI.create("http://localhost:8080"))
+                                                          .method(SdkHttpMethod.PUT)
+                                                          .build();
+
+        Context.AfterUnmarshalling afterUnmarshallingContext =
+            InterceptorTestUtils.afterUnmarshallingContext(putObjectRequest, sdkHttpRequest, response, sdkHttpResponse);
+
+        assertThatThrownBy(() -> interceptor.afterUnmarshalling(afterUnmarshallingContext, getExecutionAttributesWithChecksum()))
+                .hasMessage("Data read has a different checksum than expected.");
+    }
+
+    @Test
+    public void afterUnmarshalling_putObjectRequest_with_SSE_shouldNotValidateChecksum() {
+        SdkHttpResponse sdkHttpResponse = getSdkHttpResponseWithChecksumHeader();
+
+        PutObjectResponse response = PutObjectResponse.builder()
+                                                      .eTag(INVALID_CHECKSUM)
+                                                      .build();
+
+        PutObjectRequest putObjectRequest = PutObjectRequest.builder().build();
+
+        SdkHttpRequest sdkHttpRequest = SdkHttpFullRequest.builder()
+                                                          .putHeader(SERVER_SIDE_ENCRYPTION_HEADER, AWS_KMS.toString())
+                                                          .putHeader("x-amz-server-side-encryption-aws-kms-key-id", ENABLE_MD5_CHECKSUM_HEADER_VALUE)
+                                                          .uri(URI.create("http://localhost:8080"))
+                                                          .method(SdkHttpMethod.PUT)
+                                                          .build();
+
+        Context.AfterUnmarshalling afterUnmarshallingContext =
+            InterceptorTestUtils.afterUnmarshallingContext(putObjectRequest, sdkHttpRequest, response, sdkHttpResponse);
+
+        interceptor.afterUnmarshalling(afterUnmarshallingContext, getExecutionAttributesWithChecksum());
+    }
+
+    private ExecutionAttributes getExecutionAttributesWithChecksum() {
+        SdkChecksum checksum = checkSumFor(CONTENT_BYTES);
+        return getExecutionAttributes().putAttribute(CHECKSUM, checksum);
+    }
+
     private ExecutionAttributes getExecutionAttributes() {
         ExecutionAttributes executionAttributes = new ExecutionAttributes();
         executionAttributes.putAttribute(CLIENT_TYPE, ASYNC);
@@ -133,4 +221,11 @@ public class AsyncChecksumValidationInterceptorTest {
         executionAttributes.putAttribute(SERVICE_CONFIG, S3Configuration.builder().checksumValidationEnabled(false).build());
         return executionAttributes;
     }
+
+    private static SdkChecksum checkSumFor(byte[] bytes) {
+        SdkChecksum checksum = new Md5Checksum();
+        checksum.update(bytes, 0, bytes.length);
+        return checksum;
+    }
+
 }

--- a/services/s3/src/test/java/software/amazon/awssdk/services/s3/utils/InterceptorTestUtils.java
+++ b/services/s3/src/test/java/software/amazon/awssdk/services/s3/utils/InterceptorTestUtils.java
@@ -162,4 +162,48 @@ public final class InterceptorTestUtils {
             }
         };
     }
+
+    public static Context.AfterUnmarshalling afterUnmarshallingContext(SdkRequest request, SdkHttpRequest sdkHttpRequest, SdkResponse response, SdkHttpResponse sdkHttpResponse) {
+        return new Context.AfterUnmarshalling() {
+            @Override
+            public SdkResponse response() {
+                return response;
+            }
+
+            @Override
+            public SdkHttpResponse httpResponse() {
+                return sdkHttpResponse;
+            }
+
+            @Override
+            public Optional<Publisher<ByteBuffer>> responsePublisher() {
+                return Optional.empty();
+            }
+
+            @Override
+            public Optional<InputStream> responseBody() {
+                return Optional.empty();
+            }
+
+            @Override
+            public SdkHttpRequest httpRequest() {
+                return sdkHttpRequest;
+            }
+
+            @Override
+            public Optional<RequestBody> requestBody() {
+                return Optional.empty();
+            }
+
+            @Override
+            public Optional<AsyncRequestBody> asyncRequestBody() {
+                return Optional.empty();
+            }
+
+            @Override
+            public SdkRequest request() {
+                return request;
+            }
+        };
+    }
 }

